app/lib/components/scrollbar/Scrollbar.tsx

```diff
@@ -0,0 +1,170 @@
+import React from "react";
+import { v4 as uuid } from "uuid";
+import Disposer from "../../dispose";
+import { EventNotifier } from "../../events";
+
+export class HorizontalScrollbarController {
+  private readonly disposer = new Disposer();
+  public readonly events = new EventNotifier();
+
+  private track: HTMLDivElement | null = null;
+  private thumb: HTMLDivElement | null = null;
+
+  private scrollLeft = 0;
+  private scrollLeftMax = 0;
+
+  init(track: HTMLDivElement, scrollingElement: HTMLElement) {
+    this.track = track;
+    this.thumb = this.track.children[0] as any;
+    if (!scrollingElement.id) {
+      scrollingElement.id = uuid();
+    }
+    this.thumb!.setAttribute("aria-controls", scrollingElement.id);
+
+    this.disposer
+      .subscribe(this.thumb!, "mousedown", this.onMouseDown)
+      .subscribe(scrollingElement, "mouseenter", this.onMouseEnterScrollingElement)
+      .subscribe(scrollingElement, "mouseleave", this.onMouseOutScrollingElement)
+      .subscribe(scrollingElement, "wheel", this.onWheelScrollingElement)
+      .subscribe(window, "mousemove", this.onMouseMove)
+      .subscribe(window, "mouseup", this.onMouseUp)
+      .subscribe(window, "keydown", this.onKeyDown);
+  }
+
+  private thumbWidth = 0;
+  private scrollableWidth = 0;
+  private thumbX = 0;
+
+  update({ scrollLeft, scrollLeftMax }: { scrollLeft: number; scrollLeftMax: number }) {
+    if (!this.track) return;
+
+    this.scrollLeft = scrollLeft;
+    this.scrollLeftMax = scrollLeftMax;
+    const trackClientWidth = this.track.clientWidth;
+    this.scrollableWidth = this.scrollLeftMax + trackClientWidth;
+    this.thumbWidth = (trackClientWidth / this.scrollableWidth) * trackClientWidth;
+    this.thumbX =
+      scrollLeftMax === 0
+        ? 0
+        : (this.scrollLeft / this.scrollLeftMax) * (trackClientWidth - this.thumbWidth);
+
+    this.thumb!.style.left = `${this.thumbX}px`;
+    this.thumb!.style.width = `${this.thumbWidth}px`;
+
+    this.track.setAttribute(
+      "aria-valuenow",
+      String(
+        Math.round(
+          this.thumbX === 0 ? 0 : (100 * this.thumbX) / (trackClientWidth - this.thumbWidth)
+        )
+      )
+    );
+  }
+
+  private isScrolling = false;
+  // Single scroll event here avoids creating too many objects.
+  private scrollEvent = { delta: 0, animate: false };
+  private minMouseX = 0;
+  private maxMouseX = 0;
+  private mouseX = 0;
+  onMouseDown = (e: MouseEvent) => {
+    if (e.buttons & 1) {
+      this.isScrolling = true;
+      const thumb = this.thumb!.getBoundingClientRect();
+      const track = this.track!.getBoundingClientRect();
+      this.minMouseX = e.clientX - (thumb.x - track.x);
+      this.maxMouseX = e.clientX + (track.x + track.width - (thumb.x + thumb.width));
+    }
+    this.updateMouse(e);
+  };
+  onMouseMove = (e: MouseEvent) => {
```

Comment:
is the reason these aren't instance methods because of binding "this"?

can we make them instance methods and use this.onMouseDown.bind(this) above?

Reply:
:+1: 

Wasn't sure whether this would affect performance... did some quick benchmarking - interestingly, `.bind(this)` gets ~10X throughput if you're instantiating a lot of instances